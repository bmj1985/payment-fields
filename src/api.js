import loadjs from 'loadjs';

let NEXT_FIELD_ID = 0;

export function resetIDCounter() {
    NEXT_FIELD_ID = 0;
}

function nextFieldId() {
    NEXT_FIELD_ID += 1;
    return NEXT_FIELD_ID;
}

class Field {

    constructor(api, props) {
        this.api = api;
        this.isReady = false;
        this.events = Object.create(null);
        this.props = props;
        const { type, ...rest } = props;
        this.id = `field-wrapper-${nextFieldId()}`;
        this.type = type;
        for (const key in rest) { // eslint-disable-line
            if (key.startsWith('on')) {
                this.events[key] = rest[key];
                delete rest[key];
            }
        }
        this.options = rest;
    }

    emit(event) {
        if (this.events[event.type]) {
            this.events[event.type](event);
        }
    }

    get selector() {
        return `#${this.id}`;
    }

}


export default class ClientApi {

    static Field = Field;

    fields = Object.create(null);

    fieldHandlers = Object.create(null);

    constructor({ isReady, urls, props }) {
        const { authorization: _, styles, ...callbacks } = props;
        this.styles = styles || {};
        this.wrapperHandlers = callbacks || {};
        this.tokenize = this.tokenize.bind(this);
        if (!isReady) {
            this.fetch(urls);
        }
    }

    isApiReady() {
        return false;
    }

    fetch(urls) {
        loadjs(urls, {
            success: () => {
                if (this.pendingAuthorization) {
                    this.setAuthorization(this.pendingAuthorization);
                }
            },
            error: this.onError.bind(this),
        });
    }

    setAuthorization(authorization) {
        if (!this.isApiReady()) {
            this.pendingAuthorization = authorization;
            return;
        }
        if (!authorization && this.authorization) {
            this.teardown();
        } else if (authorization && authorization !== this.authorization) {
            if (this.authorization) { this.teardown(); }
            this.authorization = authorization;
            this.createInstance(authorization);
        }
    }

    onError(err) {
        if (!err) { return; }
        if (this.wrapperHandlers.onError) { this.wrapperHandlers.onError(err); }
    }

    checkInField(props) {
        const { FieldClass } = this;
        const field = new FieldClass(this, props);
        this.fields[field.type] = field;
        return field.id;
    }

    onFieldsReady() {
        if (this.wrapperHandlers.onReady) {
            this.wrapperHandlers.onReady({ tokenize: this.tokenize });
        }
    }

    onFieldValidity(event) {
        if (this.wrapperHandlers.onValidityChange) {
            this.wrapperHandlers.onValidityChange(event);
        }
    }

    onFieldEvent(event) {
        if (this.wrapperHandlers[event.type]) {
            this.wrapperHandlers[event.type](event);
        }
    }

    // the following fields are implemented in each client lib
    createInstance() {}

    tokenize() { }

    teardown() { }


}
