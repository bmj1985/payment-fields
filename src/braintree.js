import HostedFieldsApi from './hosted-fields-api';

// sandbox_g42y39zw_348pk9cgf3bgyw2b

export const EVENTS_MAP = {
    focus: 'onFocus',
    blur: 'onBlur',
    validityChange: 'onValidityChange',
    cardTypeChange: 'onCardTypeChange',
};

const EVENT_DECODERS = {

    onCardTypeChange(event) {
        return Object.assign(
            event.cards[0],
            { type: 'onCardTypeChange', brand: event.cards[0].type },
        );
    },

};

export const TYPES_MAP = {
    cardNumber: 'number',
};

export class BraintreeField extends HostedFieldsApi.Field {

    constructor(api, props) {
        super(api, props);

        this.type = TYPES_MAP[this.type] || this.type;

        this.options.selector = `#${this.id}`;
    }

}

export default class BraintreeApi extends HostedFieldsApi {


    FieldClass = BraintreeField;

    constructor(props) {
        super({
            props,
            isReady: !!global.braintree,
            urls: [
                'https://js.braintreegateway.com/web/3.22.2/js/client.js',
                'https://js.braintreegateway.com/web/3.22.2/js/hosted-fields.js',
            ],
        });
    }

    isApiReady() {
        return !!global.braintree;
    }

    createInstance() {
        global.braintree.client.create(
            { authorization: this.authorization },
            (err, clientInstance) => {
                if (err) {
                    this.onError(err);
                } else {
                    this.createFields(clientInstance);
                }
            },
        );
    }

    createFields(client) {
        const fields = {};
        Object.keys(this.fields).forEach((fieldName) => {
            fields[fieldName] = this.fields[fieldName].options;
        });
        global.braintree.hostedFields.create({
            client,
            fields,
            styles: {
                input: this.styles.base,
                ':focus': this.styles.focus,
                '.invalid': this.styles.invalid,
                '.valid': this.styles.valid,
            },
        }, (err, hostedFields) => {
            if (err) {
                this.onError(err);
                return;
            }
            this.hostedFields = hostedFields;
            [
                'blur', 'focus', 'empty', 'notEmpty',
                'cardTypeChange', 'validityChange',
            ].forEach((eventName) => {
                hostedFields.on(eventName, ev =>
                    this.onFieldEvent(eventName, ev));
            });
            this.onFieldsReady();
        });
    }

    onFieldEvent(eventName, event) {
        const type = EVENTS_MAP[eventName] || eventName;
        const attrs = EVENT_DECODERS[type] ? EVENT_DECODERS[type](event) : {};
        const field = this.fields[TYPES_MAP[event.emittedBy] || event.emittedBy];
        const sanitizedEvent = Object.assign({
            field: field.props.type,
            type,
            event,
        }, attrs);

        field.emit(sanitizedEvent);

        if ('validityChange' === eventName) {
            const isValid = !Object.keys(event.fields).find(f => false === event.fields[f].isValid);
            this.onFieldValidity(Object.assign(sanitizedEvent, { isValid }));
        } else {
            super.onFieldEvent(sanitizedEvent);
        }
    }

    tokenize(options = {}) {
        return new Promise((resolve, reject) => {
            this.hostedFields.tokenize(options, (err, payload) => {
                if (err) {
                    this.onError(err);
                    reject(err);
                } else {
                    resolve({ token: payload });
                }
            });
        });
    }

    teardown() {
        if (this.hostedFields) { this.hostedFields.teardown(); }
    }

}
