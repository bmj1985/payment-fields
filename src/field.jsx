import React from 'react'; // eslint-disable-line no-unused-vars
import PropTypes from 'prop-types';

export default class Field extends React.Component {

    static propTypes = {
        type: PropTypes.oneOf([
            'cardNumber', 'expirationDate', 'cvv', 'postalCode',
        ]).isRequired,
        placeholder: PropTypes.string,
        className: PropTypes.string,
        onValidityChange: PropTypes.func,
        onFocus: PropTypes.func,
        onBlur: PropTypes.func,
    }

    static defaultProps = {
        placeholder: '',
    }

    static contextTypes = {
        paymentFieldsApi: PropTypes.object,
    }

    focus() {
        this.context.paymentFieldsApi.focusField(this.props.type);
    }

    clear() {
        this.context.paymentFieldsApi.clearField(this.props.type);
    }

    componentWillMount() {
        this.fieldId = this.context.paymentFieldsApi.checkInField(this.props);
    }

    get className() {
        const list = ['payment-field'];
        if (this.props.className) { list.push(this.props.className); }
        return list.join(' ');
    }

    render() {
        return <div id={this.fieldId} className={this.className} />;
    }

}
