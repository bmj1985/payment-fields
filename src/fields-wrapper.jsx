import React from 'react';
import PropTypes from 'prop-types';

import Braintree from './braintree';
import Square from './square';
import Stripe from './stripe';

export default class CreditCardFieldsWrapper extends React.Component {

    static Vendors= {
        Stripe,
        Square,
        Braintree,
    };

    static propTypes = {
        vendor: PropTypes.string.isRequired,
        children: PropTypes.node.isRequired,
        authorization: PropTypes.string,

        onReady: PropTypes.func.isRequired,
        onValidityChange: PropTypes.func,
        onCardTypeChange: PropTypes.func,
        onError: PropTypes.func,

        passThroughStyles: PropTypes.any,
        styles: PropTypes.object,
        className: PropTypes.string,
        tagName: PropTypes.string,
    }

    static defaultProps = {
        tagName: 'div',
        styles: {},
    }

    static childContextTypes = {
        hostedFieldsApi: PropTypes.object,
    }

    constructor(props) {
        super(props);
        const Api = this.constructor.Vendors[props.vendor];
        this.api = new Api(props);
    }

    componentDidMount() {
        this.api.setAuthorization(this.props.authorization);
    }

    componentWillUnmount() {
        this.api.teardown();
    }

    componentWillReceiveProps(nextProps) {
        this.api.setAuthorization(nextProps.authorization);
    }

    tokenize(options) {
        return this.api.tokenize(options);
    }

    getChildContext() {
        return { hostedFieldsApi: this.api };
    }

    render() {
        const { className: providedClass, tagName: Tag } = this.props;
        let className = 'braintree-hosted-fields-wrapper';
        if (providedClass) { className += ` ${providedClass}`; }
        return (
            <Tag className={className}>
                {this.props.children}
            </Tag>
        );
    }

}
