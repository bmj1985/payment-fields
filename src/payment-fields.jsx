import React from 'react';
import PropTypes from 'prop-types';
import Vendors from './vendors';

import Field from './field.jsx';

export default class PaymentFields extends React.Component {

    static Field = Field;

    static propTypes = {
        vendor: PropTypes.string.isRequired,
        children: PropTypes.node.isRequired,
        authorization: PropTypes.string,

        onReady: PropTypes.func.isRequired,
        onValidityChange: PropTypes.func,
        onCardTypeChange: PropTypes.func,
        onError: PropTypes.func,

        passThroughStyles: PropTypes.any,
        styles: PropTypes.object,
        className: PropTypes.string,
        tagName: PropTypes.string,
    }

    static defaultProps = {
        tagName: 'div',
        styles: {},
    }

    static childContextTypes = {
        paymentFieldsApi: PropTypes.object,
    }

    constructor(props) {
        super(props);
        const Api = Vendors[props.vendor];
        this.api = new Api(props);
    }

    componentDidMount() {
        this.api.setAuthorization(this.props.authorization);
    }

    componentWillUnmount() {
        this.api.teardown();
    }

    componentWillReceiveProps(nextProps) {
        this.api.setAuthorization(nextProps.authorization);
    }

    tokenize(options) {
        return this.api.tokenize(options);
    }

    getChildContext() {
        return { paymentFieldsApi: this.api };
    }

    render() {
        const { className: providedClass, tagName: Tag } = this.props;
        let className = 'payment-fields-wrapper';
        if (providedClass) { className += ` ${providedClass}`; }
        return (
            <Tag className={className}>
                {this.props.children}
            </Tag>
        );
    }

}
