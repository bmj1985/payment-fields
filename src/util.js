export function capitalize(string) {
    return string.charAt(0).toUpperCase() + string.slice(1);
}

export function camelCase(str) {
    return str.replace(/-([a-z])/gi, (s, c, i) => (0 === i ? c : c.toUpperCase()));
}

export function camelCaseKeys(src) {
    if (!src) return {};
    const dest = {};
    const keys = Object.keys(src);
    for (let i = 0; i < keys.length; i++) {
        const key = keys[i];
        dest[camelCase(key)] = src[key];
    }
    return dest;
}
