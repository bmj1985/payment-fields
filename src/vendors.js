import Braintree from './braintree';
import Square from './square';
import Stripe from './stripe';

export default {
    Braintree, Square, Stripe,
};
