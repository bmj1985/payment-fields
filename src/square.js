import React from 'react'; // eslint-disable-line no-unused-vars
import HostedFieldsApi from './hosted-fields-api';
import { camelCaseKeys } from './util';
    // The Square docs note:
    // If you host a copy of this library on your own server, or use a copy hosted on a third-party server,
    // your application will be disabled without notice.
    // we have no choice but to load and pollute the global namespace.

export const EVENTS_MAP = {
    focusClassAdded: 'onFocus',
    focusClassRemoved: 'onBlur',
    errorClassAdded: 'onValidityChange',
    errorClassRemoved: 'onValidityChange',
    cardBrandChanged: 'onCardTypeChange',
    postalCodeChanged: 'onChange',
};

const EVENT_DECODERS = {

    onCardTypeChange(event) {
        return { brand: event.cardBrand };
    },

};

// sandbox-sq0idp-i06hC8ZeXrqOujH_QfYt5Q

export class SquareField extends HostedFieldsApi.Field {

    constructor(api, props) {
        super(api, props);
        this.options.elementId = this.id;
    }

}


export default class SquareApi extends HostedFieldsApi {

    FieldClass = SquareField;

    constructor(props) {
        super({
            isReady: !!global.SqPaymentForm,
            urls: ['https://js.squareup.com/v2/paymentform'],
            props,
        });
    }

    isApiReady() {
        return !!global.SqPaymentForm;
    }

    createInstance() {
        const options = {
            inputClass: 'hosted-card-field',
            applicationId: this.authorization,
            captureUncaughtExceptions: true,
            inputStyles: [
                camelCaseKeys(this.styles.base),
            ],
        };
        for (const type in this.fields) { // eslint-disable-line
            const field = this.fields[type];
            options[field.type] = field.options;
        }
        options.callbacks = {
            cardNonceResponseReceived: this.onCardNonce.bind(this),
            inputEventReceived: this.onFieldEvent.bind(this),
            paymentFormLoaded: this.onFieldsReady.bind(this),
        };
        this.hostedFields = new global.SqPaymentForm(options);
        this.hostedFields.build();
    }

    onFieldEvent(event) {
        const type = EVENTS_MAP[event.eventType] || event.eventType;
        const attrs = EVENT_DECODERS[type] ? EVENT_DECODERS[type](event) : {};

        const sanitizedEvent = Object.assign({
            field: event.field,
            type,
            event,
        }, attrs);

        this.fields[event.field].emit(sanitizedEvent);

        super.onFieldEvent(sanitizedEvent);

        if (event.currentState.isCompletelyValid !== event.previousState.isCompletelyValid) {
            this.onFieldValidity(Object.assign(sanitizedEvent, {
                type: 'validityChange',
                isValid: event.currentState.isCompletelyValid,
            }));
        }
    }

    onCardNonce(errors, nonce, cardData) {
        const { pendingToken } = this;
        this.pendingToken = null;
        if (errors) {
            pendingToken.reject({ errors });
            return;
        }
        pendingToken.resolve({ token: nonce, cardData });
    }

    tokenize() {
        if (this.pendingToken) {
            return Promise.reject({ errors: ['tokenization in progress'] });
        }
        return new Promise((resolve, reject) => {
            this.pendingToken = { resolve, reject };
            this.hostedFields.requestCardNonce();
        });
    }

    teardown() {
        if (this.hostedFields) { this.hostedFields.destroy(); }
        super.teardown();
    }

}
