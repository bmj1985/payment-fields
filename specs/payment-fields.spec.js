import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Braintree from './braintree-mocks';
import Square from './square-mocks.js';
import Stripe from './stripe-mocks.js';
import TestingComponent from './testing-component-factory.jsx';
import { resetIDCounter } from '../src/api';

const Mocks = {
    Braintree,
    Square,
    Stripe,
};

[
    'Stripe',
    'Square',
    'Braintree',
].forEach((vendor) => {
    const mountComponent = props => mount(TestingComponent.build({ vendor, ...props }));

    describe(`${vendor} Api`, () => {
        let mocks;
        const authorization = `${vendor}-123456`;

        beforeEach(() => {
            mocks = new Mocks[vendor]();
        });

        afterEach(() => {
            mocks.reset();
            resetIDCounter();
        });

        it('renders and matches snapshot', () => {
            expect(renderer.create(
                TestingComponent.build({ vendor, authorization }),
            )).toMatchSnapshot();
        });

        it('sets properties', () => {
            const styles = { base: { 'font-size': '18px' } };
            mountComponent({ authorization, styles });
            mocks.expectInitialized({ authorization, styles });
        });

        it('forwards events to fields', () => {
            const onFocus = jest.fn();
            const onBlur = jest.fn();
            const onValidityChange = jest.fn();
            const fieldTypes = ['cardNumber', 'expirationDate', 'cvv', 'postalCode'];
            const fields = {};
            fieldTypes.forEach((f) => { fields[f] = { onFocus, onBlur }; });
            mountComponent({
                authorization,
                onValidityChange,
                props: fields,
            });

            fieldTypes.forEach((f) => {
                mocks.emitFocusEvent(f);
                expect(onFocus).toHaveBeenCalledWith(
                    expect.objectContaining({
                        field: f, type: 'onFocus', isValid: false,
                    }),
                );

                mocks.emitBlurEvent(f);
                expect(onBlur).toHaveBeenCalledWith(
                    expect.objectContaining({
                        field: f, type: 'onBlur', isValid: false,
                    }),
                );
                onFocus.mockReset();
                onBlur.mockReset();
            });

            mocks.emitValid(true);
            expect(onValidityChange).toHaveBeenCalledWith(
                expect.objectContaining({ isValid: true }),
            );
            onValidityChange.mockReset();
            mocks.emitValid(false);
            expect(onValidityChange).toHaveBeenCalledWith(
                expect.objectContaining({ isValid: false }),
            );
        });

        it('can generate token', () => {
            const onReady = jest.fn();
            mountComponent({ authorization, onReady });
            expect(onReady).toHaveBeenCalledWith(expect.objectContaining({
                tokenize: expect.anything(),
            }));
            return onReady
                .mock.calls[0][0].tokenize()
                .then(token => mocks.expectValidToken(token));
        });
    });
});
