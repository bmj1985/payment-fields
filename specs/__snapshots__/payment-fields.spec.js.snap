// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Braintree Api renders and matches snapshot 1`] = `
<div
  className="payment-fields-wrapper testing-component"
>
  <div
    className="payment-field"
    id="field-wrapper-1"
  />
  <div
    className="payment-field"
    id="field-wrapper-2"
  />
  <div
    className="payment-field"
    id="field-wrapper-3"
  />
  <div
    className="payment-field"
    id="field-wrapper-4"
  />
</div>
`;

exports[`Square Api renders and matches snapshot 1`] = `
<div
  className="payment-fields-wrapper testing-component"
>
  <div
    className="payment-field"
    id="field-wrapper-1"
  />
  <div
    className="payment-field"
    id="field-wrapper-2"
  />
  <div
    className="payment-field"
    id="field-wrapper-3"
  />
  <div
    className="payment-field"
    id="field-wrapper-4"
  />
</div>
`;

exports[`Stripe Api renders and matches snapshot 1`] = `
<div
  className="payment-fields-wrapper testing-component"
>
  <div
    className="payment-field"
    id="field-wrapper-1"
  />
  <div
    className="payment-field"
    id="field-wrapper-2"
  />
  <div
    className="payment-field"
    id="field-wrapper-3"
  />
  <div
    className="payment-field"
    id="field-wrapper-4"
  />
</div>
`;
